src/otaclient/ota_core.py

```diff
@@ -786,6 +790,129 @@ def execute(self) -> None:
             shutil.rmtree(self._session_workdir, ignore_errors=True)
 
 
+class _OTAClientUpdater(_OTAUpdateOperator):
+    """The implementation of OTA client update logic."""
+
+    def __init__(
+        self,
+        client_update_control_flags: ClientUpdateControlFlags,
+        **kwargs,
+    ) -> None:
+        # ------ init base class ------ #
+        super().__init__(**kwargs)
+
+        # --- Event flag to control client update ---- #
+        self.client_update_control_flags = client_update_control_flags
+
+        # ------ setup OTA client package parser ------ #
+        self._ota_client_package = OTAClientPackage(
+            base_url=self.url_base,
+            ota_metadata=self._ota_metadata,
+            session_dir=self._session_workdir,
+        )
+
+    def _execute_client_update(self):
+        """Implementation of OTA updating."""
+        logger.info(f"execute local update({ecu_info.ecu_id=}): {self.update_version=}")
+
+        try:
+            self._handle_upper_proxy()
+            self._process_metadata()
+            self._download_client_package_resources()
+            self._wait_sub_ecus()
+            if self._is_same_client_package_version():
+                self._handle_same_version()
+            else:
+                self._perform_update()
+        except Exception as e:
+            self._handle_update_failure(e)
```

Comment:
This is core flowchart of client update. Restart will be triggered for any exceptions/errors.